ompi/mca/coll/basic/coll_basic_neighbor_allgather.c

```diff
@@ -75,13 +76,12 @@ mca_coll_basic_neighbor_allgather_cart(const void *sbuf, int scount,
                                     MCA_PML_BASE_SEND_STANDARD,
                                     comm, reqs++));
             if (OMPI_SUCCESS != rc) break;
-
-            nreqs += 2;
         }
 
         rbuf = (char *) rbuf + extent * rcount;
 
         if (MPI_PROC_NULL != drank) {
+            nreqs += 2;
```

Comment:
@bosilca Replace nreqs+=2 by nreqs++ before each request. That way, we'll be able to free exactly the right number of requests in case of failure.